### ISSUE TITLE
FusedLinearBackwardCore fallback bias gradient only reduces axis 0 — breaks rank-3+ inputs and Tensor<double>

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#234]
## Summary

`BackwardFunctions<T>.FusedLinearBackwardCore` has two paths for the
bias gradient. The BLAS fast path (line ~3397) is correct for rank-2
inputs, but the general fallback (line 3423) reduces only axis 0:

```csharp
if (inputs.Length > 2)
{
    var biasGrad = engine.ReduceSum(gradOutput, new[] { 0 }, keepDims: false);
    DifferentiableOps.AccumulateGrad(grads, inputs[2], biasGrad, engine);
}
```

For rank-3 `gradOutput` of shape `[batch, seq, features]` (standard
transformer pattern), this produces `[seq, features]` instead of
`[features]`. The bias tensor has shape `[features]` (or `[1, features]`
in callers that store biases as 2D), so the downstream Adam step fails
with `"Tensor shapes must match. Got [1, F] and [seq, F]"` when the
bias-shape vs bias-grad-shape mismatch hits `TensorAdd`.

## When the fallback fires

The BLAS fast path condition at line 3364 is:

```csharp
if (typeof(T) == typeof(float) && inputs[0].Rank == 2 && inputs[1].Rank == 2
    && BlasProvider.IsAvailable && gradOutput.IsContiguous)
```

So the broken fallback runs whenever **any** of the following is true:

1. `T` is `double` (the BLAS fast path is float-only).
2. Input has rank 3+ (common in transformers, sequence models, and
   any `[batch, seq, features]` pipeline).
3. Weights are rank 3+.
4. BLAS unavailable or gradOutput non-contiguous.

The first two cases together cover a huge fraction of real uses —
every transformer trained with `double` on `[B, T, F]` inputs hits it.

## Repro

Reproduced from the consuming AiDotNet repo's Finance smoke suite on
LagLlama<double> (and MOIRAI/UniTS/TimeGPT/TimeLLM/Timer), but the
minimal shape is:

```csharp
var input = new Tensor<double>([1, 8, 1]);         // [B, T, F_in]
var weights = new Tensor<double>([1, 24]);         // [F_in, F_out]
var bias = new Tensor<double>([1, 24]);            // [1, F_out]
using var tape = new GradientTape<double>();
var y = engine.FusedLinear(input, weights, bias, FusedActivationType.None);
                                                    // y: [1, 8, 24]
var loss = engine.TensorSum(y);
var grads = tape.ComputeGradients(loss);

// grads[bias].Shape == [8, 24]          <-- bug: should be [1, 24] (or [24])
// Expected: reduction over ALL leading axes except the feature axis
// Actual:   reduction only over axis 0, leaves seq axis in place
```

Real-world diagnostic from a LagLlama forward on `[1, 8, 1]`:

```
[ADAM-DIAG] === param/grad catalog ===
  [0]  param=[1,24]    grad=[1,1,24]   (weights — reshape rescue works)
  [1]  param=[1,24]    grad=[8,24]     ← BIAS, wrong shape, 8× elements
  [10] param=[1,1024]  grad=[8,1024]   ← another FFL bias, same pattern
  [12] param=[1,24]    grad=[8,24]     ← …
  [18] param=[1,12]    grad=[8,12]     ← …
```

Every mismatched gradient is a FFL bias, and every one of them has an
extra leading `seq_len = 8` dimension — exactly the axis-0-only
`ReduceSum` contract.

## Proposed fix

Reduce every leading non-feature axis, matching the bias tensor's
trailing dims:

```csharp
if (inputs.Length > 2)
{
    var bias = inputs[2];
    int reduceCount = gradOutput.Rank - bias.Rank + LeadingOnes(bias._shape);
    // Equivalently: reduce axes [0 .. gradOutput.Rank - bias.FeatureRank - 1]
    // where FeatureRank is the trailing rank of bias after squeezing size-1 dims.
    var axes = Enumerable.Range(0, reduceCount).ToArray();
    var biasGrad = engine.ReduceSum(gradOutput, axes, keepDims: false);
    // Reshape biasGrad back to bias._shape if bias has leading size-1 dims
    if (!biasGrad._shape.SequenceEqual(bias._shape)
        && biasGrad.Length == bias.Length)
    {
        biasGrad = engine.Reshape(biasGrad, bias._shape);
    }
    DifferentiableOps.AccumulateGrad(grads, bias, biasGrad, engine);
}
```

For PyTorch parity: `torch.nn.functional.linear`'s bias gradient is
`grad_output.reshape(-1, F).sum(0)`, which equals
"reduce all axes except the last". The same contract should hold here.

## Impact in the consuming repo

Blocks training for ~6 Finance Foundation models (LagLlama, MOIRAI,
UniTS, TimeGPT, TimeLLM, Timer) whose `FeedForwardLayer` stacks
consume rank-3 `[B, T, F]` inputs with `T = double`. Tracked alongside
the companion `BatchNormBackward` 3D issue
(#233) as the two root-cause upstream bugs
behind the `AiDotNet` repo's PR #1177 Finance smoke suite residual
failures.

---

Found while diagnosing LagLlama in `ooples/AiDotNet` PR #1177.
